prometheus/client.go

```diff
@@ -66,7 +87,147 @@ func (in *Client) GetSourceServices(namespace string, servicename string) (map[s
+	if err != nil {
+		return nil, err
+	}
+	metrics["Request count out"], err = fetchRate(api, "istio_request_count",
+		fmt.Sprintf("{source_service=\"%s.%s.%s\"}", servicename, namespace, clustername), duration, now)
+	if err != nil {
+		return nil, err
+	}
+	metrics["Request size in"], err = fetchHistogram(api, "istio_request_size",
+		fmt.Sprintf("{destination_service=\"%s.%s.%s\"}", servicename, namespace, clustername), duration, now)
+	if err != nil {
+		return nil, err
+	}
+	metrics["Request size out"], err = fetchHistogram(api, "istio_request_size",
+		fmt.Sprintf("{source_service=\"%s.%s.%s\"}", servicename, namespace, clustername), duration, now)
+	if err != nil {
+		return nil, err
+	}
+	metrics["Request duration in"], err = fetchHistogram(api, "istio_request_duration",
+		fmt.Sprintf("{destination_service=\"%s.%s.%s\"}", servicename, namespace, clustername), duration, now)
+	if err != nil {
+		return nil, err
+	}
+	metrics["Request duration out"], err = fetchHistogram(api, "istio_request_duration",
+		fmt.Sprintf("{source_service=\"%s.%s.%s\"}", servicename, namespace, clustername), duration, now)
+	if err != nil {
+		return nil, err
+	}
+	metrics["Response size in"], err = fetchHistogram(api, "istio_response_size",
+		fmt.Sprintf("{destination_service=\"%s.%s.%s\"}", servicename, namespace, clustername), duration, now)
+	if err != nil {
+		return nil, err
+	}
+	metrics["Response size out"], err = fetchHistogram(api, "istio_response_size",
+		fmt.Sprintf("{source_service=\"%s.%s.%s\"}", servicename, namespace, clustername), duration, now)
+	if err != nil {
+		return nil, err
+	}
+	metrics["Healthy replicas"], err = fetchGauge(api,
+		fmt.Sprintf("envoy_cluster_out_%s_%s_%s_http_membership_healthy", servicename, namespace, envoyClustername),
+		"", now)
+	if err != nil {
+		return nil, err
+	}
+	metrics["Total replicas"], err = fetchGauge(api,
+		fmt.Sprintf("envoy_cluster_out_%s_%s_%s_http_membership_total", servicename, namespace, envoyClustername),
+		"", now)
+	if err != nil {
+		return nil, err
+	}
+	return metrics, nil
+}
+
+func fetchGauge(api v1.API, metricName string, labels string, now time.Time) (*MetricValue, error) {
+	query := fmt.Sprintf("%s%s", metricName, labels)
+	val, err := fetchScalarDouble(api, query, now)
+	if err != nil {
+		return nil, err
+	}
+	return &MetricValue{
+		Query: query,
+		Value: val}, nil
+}
+
+func fetchRate(api v1.API, metricName string, labels string, duration string, now time.Time) (*MetricValue, error) {
+	query := fmt.Sprintf("rate(%s%s[%s])", metricName, labels, duration)
+	val, err := fetchScalarDouble(api, query, now)
+	if err != nil {
+		return nil, err
+	}
+	return &MetricValue{
+		Query: query,
+		Value: val}, nil
+}
+
+func fetchHistogram(api v1.API, metricName string, labels string, duration string, now time.Time) (*MetricHistogram, error) {
+	// Note: we may want to make returned stats configurable in the future
```

Comment:
+1, it seems Prom handles each quantile as a separate query so there is some re-work for every requested quantile.  Also, we could, I think, blast of the requested quantile queries in parallel as goroutines.